### PR TITLE
Migrate more MapLikes to Maps

### PR DESCRIPTION
**Pull request by rbuckton**
This change migrates a number of remaining `MapLike<T>` entries to `Map<T>`, as well as adds a number of additional functions for working with both `MapLike<T>` and `Map<T>`.

The changes in the compiler itself add an additional performance improvement, shaving another ~400ms off of the total emit time of Monaco and Angular.

This also disables the tslint restriction on use of the `in` operator, as offline performance tests I've run show that using `in` is faster than both `hasOwnProperty` and `obj[key]` when merely testing for the presence of a key. The `in` operator also works better when paired with `delete`, which is faster at removing entries from a `Map<T>` than setting the entry to `undefined`.

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -8475,11 +8467,11 @@ namespace ts {
...
-                    const facts = doubleEquals ?
-                        assumeTrue ? TypeFacts.EQUndefinedOrNull : TypeFacts.NEUndefinedOrNull :
-                        value.kind === SyntaxKind.NullKeyword ?
-                            assumeTrue ? TypeFacts.EQNull : TypeFacts.NENull :
-                            assumeTrue ? TypeFacts.EQUndefined : TypeFacts.NEUndefined;
+                    const facts = doubleEquals
+                        ? assumeTrue ? TypeFacts.EQUndefinedOrNull : TypeFacts.NEUndefinedOrNull
+                        : value.kind === SyntaxKind.NullKeyword
+                            ? assumeTrue ? TypeFacts.EQNull : TypeFacts.NENull
+                            : assumeTrue ? TypeFacts.EQUndefined : TypeFacts.NEUndefined;
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: Not a fan of the `?` and `:` prefix style. Would prefer to keep the existing formatting.

File: src/compiler/commandLineParser.ts [hunk 2]
```diff
@@ -486,18 +486,17 @@ namespace ts {
...
         return createCompilerDiagnostic(Diagnostics.Argument_for_0_option_must_be_Colon_1, `--${opt.name}`, namesOfType);
     }
 
     /* @internal */
     export function parseCustomTypeOption(opt: CommandLineOptionOfCustomType, value: string, errors: Diagnostic[]) {
         const key = trimString((value || "")).toLowerCase();
         const map = opt.type;
-        if (hasProperty(map, key)) {
+        if (key in map) {
             return map[key];
```
Comment (ahejlsberg) [inline src/compiler/commandLineParser.ts]: Love `in` instead of `hasProperty`.

File: src/compiler/core.ts [hunk 3]
```diff
@@ -350,80 +367,271 @@ namespace ts {
...
+    /**
+     * Counts the properties of a map.
+     *
+     * NOTE: This is intended for use with Map<T> objects. For MapLike<T> objects, use
+     *       countOwnProperties instead as it offers better runtime safety.
+     *
+     * @param map A map whose properties should be counted.
+     * @param predicate An optional callback used to limit which properties should be counted.
+     */
+    export function countProperties<T>(map: Map<T>, predicate?: (value: T, key: string) => boolean) {
```
Comment (ahejlsberg) [inline src/compiler/core.ts]: Are all of these new helper functions actually used somewhere? Just making sure we don't carry around dead weight.

**Comment by ahejlsberg** [comment]
:+1: Once comments are addressed.

Thanks for cleaning all of this up. Much nicer with these changes. Just confirming that we still use `MapLike<T>` for the compiler APIs so minimize breaking changes?

File: src/compiler/checker.ts (hunk 1 above)
Comment (rbuckton) [inline src/compiler/checker.ts]: I can change it back. I primarily made the change to work around an issue with our .tmLanguage syntax coloring as the existing formatting resulted in syntax coloring being disabled/incorrect for the next 30 lines or so in VS Code.

File: src/compiler/core.ts (hunk 3 above)
Comment (rbuckton) [inline src/compiler/core.ts]: I will take another look around, but generally I was adding them to consolidate functions we use in various places (some from utilities, and some commonly used in some of our tests). I can move the ones only used in tests off to somewhere in harness if necessary.

**Comment by rbuckton** [comment]
I left everything as `MapLike<T>` that was not marked `/* @internal */`, so there should be no breaking changes against our public API.

File: src/compiler/checker.ts (hunk 1 above)
Comment (bterlson) [inline src/compiler/checker.ts]: Both are nigh impossible to parse for a human, fwiw :-P

File: src/compiler/commandLineParser.ts (hunk 2 above)
Comment (bterlson) [inline src/compiler/commandLineParser.ts]: `in` will do a proto walk. Make sure your maps have a null prototype, or else `'toString' in map` will always be true.

File: src/compiler/checker.ts (hunk 1 above)
Comment (rbuckton) [inline src/compiler/checker.ts]: @bterlson, so how's that pattern matching proposal coming along? :)

File: src/compiler/commandLineParser.ts (hunk 2 above)
Comment (rbuckton) [inline src/compiler/commandLineParser.ts]: @bterlson that's the point. Anything that is a `Map<T>` in the compiler now will always have a `null` prototype.

**Pull request merged**